File: project/src/create-chart.ts
import fetch from 'node-fetch'

import { CreateChartInput as Input, CreateChartOutput as Output } from './shared.ts'

export async function main(input: Input): Promise<Output> {
  const res = await fetch('https://quickchart.io/chart/create', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      backgroundColor: 'transparent',
      width: 720,
      height: 480,
      format: 'png',
      chart: {
        type: 'bar',
        data: input,
      },
    }),
  })

  const { url } = await res.json();
  return { url: url }
}

if (require.main === module) {
  const json = process.argv[2]

  if (json) {
    main(JSON.parse(json))
      .then(response => console.log(response))
  }
}

File: project/src/create_chart.rs
use anyhow::Result;
use reqwest::Client;
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::collections::HashMap;
use std::env;
use std::io;

#[derive(Debug, Deserialize, Serialize)]
struct Dataset {
  label: String,
  data: Vec<f32>,
}

#[derive(Debug, Deserialize, Serialize)]
struct Input {
  labels: Vec<String>,
  datasets: Vec<Dataset>,
}

async fn create_chart(params: Value) -> Result<()> {
    let input: Input = serde_json::from_value(params)?;

    let body = json!({
      "backgroundColor": "transparent",
      "width": 720,
      "height": 480,
      "format": "png",
      "chart": {
        "type": "bar",
        "data": input,
      }
    });

    let client = Client::new();

    client
        .post("https://quickchart.io/chart/create")
        .json(&body)
        .send()
        .await?;

    Ok(())
}

#[tokio::main]
async fn main() -> Result<()> {
    let params: Value = serde_json::from_str(&env::args().nth(1).expect("no argument specified"))?;

    match create_chart(params).await {
        Ok(output) => serde_json::to_writer(io::stdout(), &output)?,
        Err(err) => serde_json::to_writer(io::stdout(), &json!({"error": err.to_string()}))?,
    }

    Ok(())
}
